File: src/lib/kafe/src/main/java/kafe/BaseDescribe.java
package kafe;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

/**
 * BaseDescribe extends BaseSetup for Describe-It pattern.
 * Describe can be nested, and Its can mix mutations and assertions.
 */
public abstract class BaseDescribe<I, S, R> extends BaseSetup<I, S, R> {
    public List<BaseIt<R>> its;
    
    public BaseDescribe(
        List<String> features,
        List<BaseIt<R>> its,
        Function<S, R> describeCb,
        Object initialValues
    ) {
        // For Describe-It, we'll process its differently
        super(
            features,
            new ArrayList<>(),
            new ArrayList<>(),
            describeCb,
            initialValues
        );
        this.its = its != null ? its : new ArrayList<>();
    }
    
    // Alias setupThat to describeThat for Describe-It pattern
    @Override
    public abstract R setupThat(
        S subject,
        ITTestResourceConfiguration testResourceConfiguration,
        Object artifactory,
        Function<S, R> setupCB,
        Object initialValues
    );
    
    // Alias setup to describe for Describe-It pattern
    public R describe(
        S subject,
        String key,
        ITTestResourceConfiguration testResourceConfiguration,
        Function<Object, Boolean> tester,
        Object artifactory,
        int suiteNdx
    ) throws Exception {
        return super.setup(subject, key, testResourceConfiguration, tester, artifactory, suiteNdx);
    }
}

File: src/lib/pitono/src/base_describe.py
from .base_setup import BaseSetup
from typing import List, Any, Callable, Optional
from .pitono_types import ITestResourceConfiguration, ITestArtifactory

class BaseDescribe(BaseSetup):
    def __init__(
        self,
        features: List[str],
        its: List[Any],
        describe_cb: Callable,
        initial_values: Any
    ):
        # Map its to actions (they can mix mutations and assertions)
        super().__init__(
            features,
            its,  # Its are treated as actions
            [],   # No separate checks for Describe-It pattern
            describe_cb,
            initial_values
        )
        self.its = its
    
    async def setup_that(
        self,
        subject: Any,
        test_resource_configuration: ITestResourceConfiguration,
        artifactory: ITestArtifactory,
        setup_cb: Callable,
        initial_values: Any
    ) -> Any:
        # Default implementation for Describe setup
        if callable(setup_cb):
            return setup_cb(subject, initial_values)
        return subject
    
    # Alias setup to describe for Describe-It pattern
    async def describe(
        self,
        subject: Any,
        key: str,
        test_resource_configuration: ITestResourceConfiguration,
        tester: Callable[[Any], bool],
        artifactory: Optional[ITestArtifactory] = None,
        suite_ndx: Optional[int] = None
    ) -> Any:
        return await super().setup(
            subject,
            key,
            test_resource_configuration,
            tester,
            artifactory,
            suite_ndx
        )
